_posts/2021-01-05-overlappable-instances.markdown

```diff
@@ -0,0 +1,263 @@
+---
+layout: post
+title:  "Overlappable instances"
+date:   2021-01-05 00:00:12 +0100
+categories: haskell effects
+---
+There are many patterns, styles and libraries which can be chosen for dealing with effects in Haskell. Today I'll try to bring closer one of them.
+
+Let's start off with a very simple program written using tagless final style.
+The program inserts a value "1" into the store under the key "key1" and then tries to retrieve it using operations defined in the `Store` type class. It logs what is happening while doing the above thanks to the `Logger`.
+The instances needed for `IO` are added only to make the whole thing compile, the implementations aren't very useful (for now!)
+
+
+```haskell
+{-# LANGUAGE DerivingStrategies,UndecidableInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving, FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses, FlexibleInstances #-}
+module Effects.Overlapping where
+
+import Prelude hiding (log)
+
+class Store m where
+  put :: k -> a -> m ()
+  get :: k -> m (Maybe a)
+
+class Logger m where
+  log :: msg -> m ()
+
+program ::
+     Monad m
+  => Logger m
+  => Store m
+  => m ()
+program = do
+  put key1 val1
+  log $ "Inserted value: " <> show val1 <> " under key: " <> key1
+  maybeV <- get key1
+  case maybeV of
+    Just v -> log $ "Retrieved: " <> v
+    Nothing -> log "No data found"
+  where
+    key1 = "key1"
+    val1 :: Integer
+    val1 = 1
+
+instance Store IO where
+  put _ _ = pure ()
+  get _ = pure Nothing
+
+instance Logger IO where
+  log _ = pure ()
+
+main :: IO ()
+main = program
+```
+This is clearly limiting, because we are allowed to have only one typeclass instance for a given monad. If we would like to have a NoOp instance and a real-world instance, we have to use newtype wrappers. Let's add aliases then for our not-so-useful instances like this:
```

Comment:
nitpick: "we are allowed to have only one typeclass instance for a given monad **transformer**."

Reply:
> nitpick: "we are allowed to have only one typeclass instance for a given monad **transformer**."

ooo hi @akrmn :D